### PR TITLE
fix: missing logs on pytest failures

### PR DESCRIPTION
**Pull request by kostasrim**
On `failed assertions` pytest would sometimes copy the logs to the failed folder `before` the instance is actually `stopped`. To further complicate things it turns out that this was a bigger issue because we copied the failures per session but sometimes certain fixtures had `session` scope (for example `df_server`) which would always cause loss of logs because the instances would `stop right after` we copied their logs to the tmp folder. 

The solution is hacky but it works. We add another indirection to keep track of the failed logs for each test. Each time a test fails, we announce that the test failed and the `name of the logs that failed` and we copy their path to another file which at the `end of the session` copy back to the `/tmp/failed` and upload the logs. This works well for `session` fixtures as well, providing the write clean up semantics.

P.s. SetUp and Teardown evaluation is a little bit of a mess in python and it's clear to me why people discourage it. Unfortunately for our use case it's required.

### CODE REVIEW

**Comment by kostasrim** [comment]
https://github.com/dragonflydb/dragonfly/actions/runs/9681419582/job/26711964457?pr=3226 all good tested in various setups

**Comment by kostasrim** [comment]
@adiholden there is one more corner case I will ping when to review

File: tests/dragonfly/conftest.py [hunk 1]
```diff
@@ -334,38 +337,80 @@ def with_ca_tls_client_args(with_tls_client_args, with_tls_ca_cert_args):
...
+
+# The only issue here is that this also clears the logs for session wide fixtures which
+# could be problematic for tests that failed way later in the session. Example:
+# suppose a session fixture like dragonfly_factory run by 3 individual tests
+# First one instantiates the dragonfly instance and writes its log path in the
+# last_test_log_files.txt. Second test then clears this, and third one fails without any logs.
+# This problem existed before but luckilly so far hasn't come up. There is no really good fix for
+# this as it's hard to track which test used which fixture when it failed, since we only interact
+# with the failed test via the pytest hook pytest_exception_interact.
+@pytest.fixture(autouse=True, scope="function")
```
Comment (kostasrim) [inline tests/dragonfly/conftest.py:410]: We can see this cascading here: https://github.com/dragonflydb/dragonfly/actions/runs/9681657533/job/26712730878?pr=3226#step:12:905

There are two solutions for this:

1. Keep a global list of session wide fixtures and their log paths. Upload them all when a test case fails but we will have to manually find them which means looking at the CI the port and grepping the big list of uploaded files
2. Track down where the same instance/fixture is used multiple times and either replace a session fixture with a function fixture and see if this increases total time of tests.
3. We can create a per function fixture that polls the session dfly instances and temporary append them to the path. That way the local failure will contain the interesting logs

I am leaning towards 3 because it can solve this nicely. I will play around but on a different PR. It's an issue on it's own that existed before this PR.

**Comment by kostasrim** [comment]
new test failure https://github.com/dragonflydb/dragonfly/actions/runs/9681657533/job/26712730878?pr=3226#step:12:905 

TODO create an issue

File: tests/dragonfly/conftest.py [hunk 2]
```diff
@@ -334,38 +337,80 @@ def with_ca_tls_client_args(with_tls_client_args, with_tls_ca_cert_args):
...
+# at the end respecting the evaluation order mentioned above
+@pytest.fixture(autouse=True, scope="session")
 def run_before_and_after_test():
-    # Setup: logic before any of the test starts
-    # Empty the log on each run
-    last_log_file = open("/tmp/last_test_log_files.txt", "w")
-    last_log_file.close()
+    logging.info("Session start for run_before_and_after_test")
+    # Setup: at the start of the session
+    last_log_file = open("/tmp/last_test_log_files.txt", "w").close()
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:390]: why do we need /tmp/last_test_log_files.txt ?
can we check if the test failed when stopping the dragonfly process and if so copy its logs to /tmp/failed using self.log_files?

File: tests/dragonfly/conftest.py [hunk 3]
```diff
@@ -334,38 +337,80 @@ def with_ca_tls_client_args(with_tls_client_args, with_tls_ca_cert_args):
...
-def pytest_exception_interact(node, call, report):
-    if report.failed:
-        copy_failed_logs_and_clean_tmp_folder(report)
-
 
-@pytest.fixture(autouse=True)
+        # Clean it
+        last_log_file = open("/tmp/last_test_log_files.txt", "w").close()
+        global TEST_FAILED
+        TEST_FAILED = True
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:373]: so we need a global TEST_FAILED or maybe we can use instead python hook pytest_runtest_makereport

File: tests/dragonfly/conftest.py (hunk 2 above)
Comment (kostasrim) [reply tests/dragonfly/conftest.py:390]: > can we check if the test failed when stopping the dragonfly process && why do we need /tmp/last_test_log_files.txt ?

We can check if the test failed when stopping the instance but we loose the information `which test failed with what logs`. Basically what you suggest, would almost work reasonably well for df instances with `function` scope (it would break if within a test we did `start -> stop -> start` because we would loose logs, since the stop could succeed and since there is no failure we end up copying nothing only for df to crash after the second start and we just lost the first log here) but would spectacularly `break` for df instances/fixtures with `session scope`. Why? Because, every time we run a test, we store the name of the logs for the instances we create in `last_test_log_files` so when a test fails, we print also the `name of the logs` for that test(e.g,: https://github.com/dragonflydb/dragonfly/actions/runs/9708171685/job/26794538269#step:6:438). The number one problem of copying the logs when we `stop` is that for `session fixtures` we simply don't know when they were created and because of that we will upload the logs at the end of the session but we won't know when the test failed which logs to associate with it (because it can be that multiple logs are uploaded from different fixtures and we need to know at the point of the test failure what are the `name of the logs`. 

I do have an idea which I want to try. I want to wrap up something first and will get back very soon on this

File: tests/dragonfly/conftest.py (hunk 3 above)
Comment (kostasrim) [reply tests/dragonfly/conftest.py:373]: `python hook pytest_runtest_makereport` This is per test, `TEST_FAILED` is used at the end of the test session. Would you mind expanding your idea on how `hook pytest_runtest_makereport` would help here?

File: tests/dragonfly/conftest.py (hunk 1 above)
Comment (kostasrim) [reply tests/dragonfly/conftest.py:455]: fixed on the last commit

File: tests/dragonfly/conftest.py (hunk 2 above)
Comment (kostasrim) [reply tests/dragonfly/conftest.py:397]: yep I think now we cover all the cases and we also don't need a separate PR. I summarized in comments what is being done and why :)
